### PR TITLE
Blog app - controllers specs 

### PR DESCRIPTION
**Pull request by ZabihullahNooriWardak**
In this project, I have fulfilled the following requirements:

- Added RSpec as a gem to the project.
- Created a Request spec file for all controllers.
- Ensured that for each controller action, I checked the following:
- If the response status was correct.
- If the correct template was rendered.
- If the response body included the correct placeholder text

### CODE REVIEW

File: spec/requests/users_spec.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    expect(response.body).to match(/#{template}/)
+  end
+
+  it 'includes the correct placeholder' do
+    get '/users/1'
+    puts response.status
+    placeholder = 'Single user page'
+    expect(response.body).to match(/#{placeholder}/)
+  end
+end
```
Comment (mckent05) [inline spec/requests/users_spec.rb:45]: - You have done well so far, but some of your tests are not passing, please ensure all tests pass before requesting another review. You can consider the following tips to help you debug the issues:
   - There is an error in your `post#index` and `user#index` view files. Consider replacing all logic there with text placeholders
   - Also, your `user#show` and `post#show` actions in your controllers are causing errors. Please fix this 🙏 
vhjvhjv

- Please remove all print statements from all your test files, this is not good practice, please fix this 👍 👍

File: spec/requests/users_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    puts response.status
+    placeholder = 'All users page'
+    expect(response.body).to match(/#{placeholder}/)
+  end
+end
+
+describe 'GET /show' do
+  it 'returns a successful response' do
+    get '/users/1'
+    puts response.status
```
Comment (mckent05) [inline spec/requests/users_spec.rb:28]: - Please remove all print statements from all your test files, this is not good practice, please fix this 👍 👍 
**OPTIONAL** - Please add the gem `rails-controller-testing` to your gem file, This gem brings back assigns to your controller tests as well as assert_template to both controller and integration tests. Please fix this 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @ZabihullahNooriWardak, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍
- Good job adding tests for your controllers 👍 👍 


## Required Changes ♻️
- Please ensure all your tests pass before requesting another review 😄 
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/requests/posts_spec.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    template = File.read('app/views/posts/show.html.erb')
+    expect(response.body).to match(/#{template}/)
+  end
+
+  it 'includes the correct placeholder' do
+    get '/users/1/posts/1'
+    placeholder = 'Single post page for a given user'
+    expect(response.body).to match(/#{placeholder}/)
+  end
+end
```
Comment (skyv26) [inline spec/requests/posts_spec.rb:40]: - [ ] Amazing @ZabihullahNooriWardak, I can see the hard efforts that you put into this milestone to embrace your knowledge in testing your back-end application thoroughly. But, after careful consideration of your implementation with the senior code reviewer, I would like to suggest you, please adhere to the project requirement and try to implement the test in the right way, a way which is used by the companies as per the standard and community guidelines. You used the file system for matching the snapshot of your page which is OK but not the right way of testing your endpoints functionalities,  let me share some pointers on what you did and what is possible or not with your approach:

    1. **Usage of File.read**: Reading the template file using File. read and then attempting to match it with the response body may not be the most effective approach. The response body can include additional information, such as dynamic content or formatting, which may not match the template file exactly.

    2. **Using regular expression (match)**: Matching the response body with the template using a regular expression might lead to false positives or negatives. It's generally better to use more specific assertions.
    
**Please you can use this [guide](https://www.microverse.org/blog/understanding-rspec-controller-request-specs-integration-test-using-capybara) for a better understanding regarding the implementation of the test for the controller**

**Review by skyv26 — CHANGES_REQUESTED** [review]
## Hi @ZabihullahNooriWardak, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## STATUS: REQUIRE CHANGES 🔴 



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implement the features ✅ 

## Required Changes ♻️

- [ ] Please make sure that after implementing the changes that I asked in my `inline` comments your test should pass completely without any error. I have some guidelines for better understanding the tests implementation in ruby on rails.

Capybara is a testing tool for web applications, and it's often used in the Ruby on Rails framework for feature testing and integration testing. It provides a high-level API for interacting with web pages, simulating user interactions, and making assertions about the behavior of the application.

In Rails testing, you often have different types of tests to cover different aspects of your application:

1. **Type: :request:** These tests are also known as request specs. They focus on testing the application's behavior from the user's perspective, simulating requests and responses. They are typically used to test the entire stack – from the controller down to the database.

   Example:
   ```ruby
   describe "User authentication" do
     it "signs in the user" do
       post login_path, params: { email: "user@example.com", password: "password" }
       expect(response).to redirect_to(dashboard_path)
     end
   end
   ```

2. **Type: :controller:** These tests focus specifically on the controllers in your Rails application. They are used to test the behavior of controller actions, ensuring that the correct response is generated based on the input.

   Example:
   ```ruby
   describe UsersController do
     it "renders the new template" do
       get :new
       expect(response).to render_template(:new)
     end
   end
   ```

3. **Type: :model:** These tests are used to test the behavior of your models. They ensure that the models' validations, associations, and other business logic are functioning as expected.

   Example:
   ```ruby
   describe User do
     it "is invalid without an email" do
       user = User.new(email: nil)
       expect(user).to be_invalid
     end
   end
   ```

Regarding Selenium WebDriver and controller spec testing, Selenium is typically used for feature tests or integration tests that involve interacting with the application through a browser. In controller spec testing, the focus is on the controller's behavior, and you usually don't interact with the application through a browser. Instead, controller tests generally simulate HTTP requests and check the resulting HTTP responses.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

#### Cheers, and Happy coding!👏👏👏

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Nemwel-Boniface — COMMENTED** [review]
# Hello @ZabihullahNooriWardak  :wave:

While to understand that the changes that were requested are "new" and were not really your fault but rather as a result of an error one of the reviewers made (The first reviewer ought to have requested them), I would like you to kindly tag me @Nemwel-Boniface once you are done making the changes that were requested by @skyv26   and I will perform a manual review on your project.

You do not have to request another review once you are done with making the changes that are requested.

Happy coding :rocket:

**Pull request merged**